presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,235 @@
+        verify(client != null);
+        try {
+            GetMappingsRequest mappingsRequest = new GetMappingsRequest().types(tableDescription.getType());
+            String index = tableDescription.getIndex();
+
+            if (!isNullOrEmpty(index)) {
+                mappingsRequest.indices(index);
+            }
+            ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = client.admin().indices().getMappings(mappingsRequest).actionGet(timeout.toMillis()).getMappings();
+
+            Iterator<String> indexIterator = mappings.keysIt();
+            while (indexIterator.hasNext()) {
+                MappingMetaData mappingMetaData = mappings.get(indexIterator.next()).get(tableDescription.getType());
+                JSONObject mappingObject = new JSONObject(mappingMetaData.source().toString()).getJSONObject(tableDescription.getType());
+                JSONObject properties = mappingObject.getJSONObject("properties");
+                List<String> lists = new ArrayList<String>();
+                if (properties.has("_meta")) {
+                    JSONObject meta = mappingObject.getJSONObject("_meta");
+                    if (meta.has("lists")) {
+                        JSONArray arrays = meta.getJSONArray("lists");
+                        for (int i = 0; i < arrays.length(); i++) {
+                            lists.add(arrays.getString(i));
+                        }
+                    }
+                }
+
+                for (String columnMetadata : getColumnsMetadata(null, properties)) {
+                    Optional<ElasticsearchColumn> column = createColumn(columnMetadata, lists);
+                    if (!column.isPresent()) {
+                        continue;
+                    }
+                    if (columns.stream().noneMatch(col -> col.getName().equals(column.get().getName()))) {
+                        columns.add(column.get());
+                    }
+                }
+            }
+            columns.add(createColumn("_id.type:string", ImmutableList.of()).get());
+            columns.add(createColumn("_index.type:string", ImmutableList.of()).get());
+            tableDescription.setColumns(ImmutableList.copyOf(columns));
+            tableDescription.setColumnsMetadata(columns.stream().map(ElasticsearchColumnMetadata::new).collect(Collectors.toList()));
+        }
+        catch (JSONException e) {
+            throw new PrestoException(ELASTIC_SEARCH_MAPPING_REQUEST_ERROR, e);
+        }
+    }
+
+    private List<String> getColumnsMetadata(String parent, JSONObject json)
+            throws JSONException
+    {
+        List<String> leaves = new ArrayList();
+        Iterator it = json.keys();
+        while (it.hasNext()) {
+            String key = (String) it.next();
+            Object child = json.get(key);
+            String childKey = parent == null || parent.isEmpty() ? key : parent.concat(".").concat(key);
+
+            if (child instanceof JSONObject) {
```

Comment:
A condition with just a `continue` looks weird (in this case it's not necessary either, so it can be an empty block). How about:

```
if (child instanceof JSONObject) {
    leaves.addAll(getColumnsMetadata(childKey, (JSONObject) child));
    continue;
}

if (!(child instanceof JSONArray)) {
    leaves.add(childKey.concat(":").concat(child.toString()));
}
```